eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs

```diff
@@ -0,0 +1,971 @@
+      , opt (idx 11 ==> unit_interval) -- treasury growth rate
+      , opt (idx 16 ==> coin) -- min pool cost
+      , opt (idx 17 ==> coin) -- ada per utxo byte
+      , opt (idx 18 ==> costmdls) -- cost models for script languages
+      , opt (idx 19 ==> ex_unit_prices) -- execution costs
+      , opt (idx 20 ==> ex_units) -- max tx ex units
+      , opt (idx 21 ==> ex_units) -- max block ex units
+      , opt (idx 22 ==> (VUInt `sized` (4 :: Word64))) -- max value size
+      , opt (idx 23 ==> (VUInt `sized` (2 :: Word64))) -- collateral percentage
+      , opt (idx 24 ==> (VUInt `sized` (2 :: Word64))) -- max collateral inputs
+      , opt (idx 25 ==> pool_voting_thresholds) -- pool voting thresholds
+      , opt (idx 26 ==> drep_voting_thresholds) -- DRep voting thresholds
+      , opt (idx 27 ==> (VUInt `sized` (2 :: Word64))) -- min committee size
+      , opt (idx 28 ==> epoch) -- committee term limit
+      , opt (idx 29 ==> epoch) -- governance action validity period
+      , opt (idx 30 ==> coin) -- governance action deposit
+      , opt (idx 31 ==> coin) -- DRep deposit
+      , opt (idx 32 ==> epoch) -- DRep inactivity period
+      ]
+
+pool_voting_thresholds :: Rule
+pool_voting_thresholds =
+  "pool_voting_thresholds"
+    =:= arr
+      [ a unit_interval -- motion no confidence
+      , a unit_interval -- committee normal
+      , a unit_interval -- committee no confidence
+      , a unit_interval -- hard fork initiation
+      , a unit_interval -- security relevant parameter voting threshold
+      ]
+
+drep_voting_thresholds :: Rule
+drep_voting_thresholds =
+  "drep_voting_thresholds"
+    =:= arr
+      [ a unit_interval -- motion no confidence
+      , a unit_interval -- committee normal
+      , a unit_interval -- committee no confidence
+      , a unit_interval -- update constitution
+      , a unit_interval -- hard fork initiation
+      , a unit_interval -- PP network group
+      , a unit_interval -- PP economic group
+      , a unit_interval -- PP technical group
+      , a unit_interval -- PP governance group
+      , a unit_interval -- treasury withdrawal
+      ]
+
+transaction_witness_set :: Rule
+transaction_witness_set =
+  "transaction_witness_set"
+    =:= mp
+      [ opt $ idx 0 ==> nonempty_set vkeywitness
+      , opt $ idx 1 ==> nonempty_set native_script
+      , opt $ idx 2 ==> nonempty_set bootstrap_witness
+      , opt $ idx 3 ==> nonempty_set plutus_v1_script
+      , opt $ idx 4 ==> nonempty_set plutus_data
+      , opt $ idx 5 ==> redeemers
+      , opt $ idx 6 ==> nonempty_set plutus_v2_script
+      , opt $ idx 7 ==> nonempty_set plutus_v3_script
+      ]
+
+-- The real type of  plutus_v1_script, plutus_v2_script and plutus_v3_script is bytes.
+-- However, because we enforce uniqueness when many scripts are supplied,
+-- we need to hack around for tests in order to avoid generating duplicates,
+-- since the cddl tool we use for roundtrip testing doesn't generate distinct collections.
```

Comment:
Is this comment still applicable? it seems like `distinct` has solved the problem that's being described in, in which case we can get rid of it.
```suggestion
```

Reply:
Isn't `distinct` a hack, though? Precisely the hack that we're referring to.

Reply:
Sorry, I assumed we don't need the hack anymore, but indeed, since we still use distinct in the type itself, the comment makes sense.